datafusion/physical-plan/src/aggregates/row_hash.rs

```diff
@@ -609,15 +579,56 @@ impl Stream for GroupedHashAggregateStream {
             match &self.exec_state {
                 ExecutionState::ReadingInput => 'reading_input: {
                     match ready!(self.input.poll_next_unpin(cx)) {
+                        // New batch to aggregate in partial aggregation operator with skip aggregation probe enabled
+                        Some(Ok(batch))
+                            if self.mode == AggregateMode::Partial
+                                && self.skip_aggregation_probe.is_some() =>
+                        {
+                            let timer = elapsed_compute.timer();
+
+                            // Do the grouping
+                            extract_ok!(
+                                self.group_aggregate_batch_with_skipping_partial(&batch)
+                            );
+                            if self.skip_partial_aggregation {
+                                let states = self.transform_to_states(batch)?;
+                                self.exec_state = ExecutionState::ProducingOutput(states);
```

Comment:
> And it seems that just the last batch (batch n in the exmaple)'s updating can be avoided, comparing to the original partial skipping logic. And the cost of updating one batch may be not so high?

Why only the last batch is not computed? If batch m reach the skip threshold, I think m to n are all skipped, therefore n-m computation are saved.

Reply:
> > And it seems that just the last batch (batch n in the exmaple)'s updating can be avoided, comparing to the original partial skipping logic. And the cost of updating one batch may be not so high?
> 
> Why only the last batch is not computed? If batch m reach the skip threshold, I think m to n are all skipped, therefore n-m computation are saved.

Oh, I think maybe I got it? Do you mean the loop here?
```
        for (index, group_values) in group_by_values.iter().enumerate() {
              ...
              reach the threshold, and we break the loop and skip earlier
              ...
        }
```

Reply:
Not really, for simple group by case, `group_by_values` is mostly single, I think only grouping sets query have more than one group by values.

Reply:
```rust
If any batch reach the threshold limit, we will switch to ProducingOutput

if self.skip_partial_aggregation {
      let states = self.transform_to_states(batch)?;
      self.exec_state = ExecutionState::ProducingOutput(states);
      // make sure the exec_state just set is not overwritten below
      break 'reading_input;
  }
```

I think the skipping logic has no difference from the `main` branch, the only difference is that we compute the hash, and skip based on the number of unique value in hash table

Reply:
> Not really, for simple group by case, `group_by_values` is mostly single, I think only grouping sets query have more than one group by values.

🤔  Yes, I am just a bit confused about `saving n - m computation` in https://github.com/apache/datafusion/pull/12697#discussion_r1788514197 if it is said single group by case.

Assume the threshold is reached until `batch n` comes:
- In main, we will firstly update `batch n` into `GroupValues` and `GroupAccumulator`s, and found threshold reached after, so we skipped `batch n+1 ~ batch last`
``` rust
// `batch n` comes, update it into `GroupValues` and `GroupAccumulator`s.
extract_ok!(self.group_aggregate_batch(batch));

// Then we found threshold reached here after updating.
self.update_skip_aggregation_probe(input_rows);

// Switch to skip mode, and we skip `batch n+1 ~ batch last`
extract_ok!(self.switch_to_skip_aggregation());
``` 

- And currently, we can know threshold reached before updating `batch n` into `GroupValues` and `GroupAccumulator`s, so we can skip `batch n ~ batch last`
``` rust
fn group_aggregate_batch_with_skipping_partial(
    &mut self,
    batch: &RecordBatch,
) -> Result<()> {
   for (index, group_values) in group_by_values.iter().enumerate() {
         // We calculate hashes and check skip threshold here
         // When `batch n` comes, we found threshold reached, 
         // just return and not update `batch n`.
         
         // And it will switch to skip mode for `batch n ~ batch last` after return    
   } 
    
   // We will still update `batch 0 ~ batch n-1`, because threshold is only
   // reached until `batch n` comes.
   for (index, group_values) in group_by_values.iter().enumerate() {
          // Update `batch` into `GroupValues` and `GroupAccumulator`s
    }
  
``` 

Seems calculation of `batch n` is avoided?

😄 Sorry for maybe asking too many questions.

Reply:
Oh, I see. It tells why the improvement is not as high as expected 🤔 

I think the attempt for this change failed 😢

Reply:
😢